src/Tizen.Applications.ComponentManager/Tizen.Applications/ComponentManager.cs

```diff
@@ -0,0 +1,186 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Tizen.Applications
+{
+    /// <summary>
+    /// This class has the methods and events of the ComponentManager.
+    /// </summary>
+    /// <since_tizen> 6 </since_tizen>
+    public static class ComponentManager
+    {
+        private const string LogTag = "Tizen.Applications";
+        
+        /// <summary>
+        /// Gets the information of the installed components asynchronously.
+        /// </summary>
+        /// <returns>The installed component info list.</returns>
+        /// <exception cref="ArgumentException">Thrown when failed because of an invalid argument.</exception>
+        /// <exception cref="InvalidOperationException">Thrown when failed because of the "component not exist" error or the system error.</exception>
+        /// <exception cref="OutOfMemoryException">Thrown when failed because of ouf of memory.</exception>
+        /// <exception cref="UnauthorizedAccessException">Thrown when failed because of permission denied.</exception>
+        /// <privilege>http://tizen.org/privilege/packagemanager.info</privilege>
+        /// <since_tizen> 6 </since_tizen>
+        public static async Task<IEnumerable<ComponentInfo>> GetInstalledComponentsAsync()
+        {
+            return await Task.Run(() =>
+            {
+                Interop.ComponentManager.ErrorCode err = Interop.ComponentManager.ErrorCode.None;
+                List<ComponentInfo> result = new List<ComponentInfo>();
+
+                Interop.ComponentManager.ComponentManagerComponentInfoCallback cb = (IntPtr infoHandle, IntPtr userData) =>
+                {
+                    if (infoHandle != IntPtr.Zero)
+                    {
+                        IntPtr clonedHandle = IntPtr.Zero;
+                        err = Interop.ComponentManager.ComponentInfoClone(out clonedHandle, infoHandle);
+                        if (err != Interop.ComponentManager.ErrorCode.None)
+                        {
+                            Log.Warn(LogTag, "Failed to clone the ComponentInfo. err = " + err);
+                            return false;
+                        }
+                        ComponentInfo info = new ComponentInfo(clonedHandle);
+                        result.Add(info);
+                        return true;
+                    }
+                    return false;
+                };
+                err = Interop.ComponentManager.ComponentManagerForeachComponentInfo(cb, IntPtr.Zero);
```

Comment:
Is it safe for thread?

Reply:
``` c

1217static int __get_comp_info_list(uid_t uid, GList **info_list)
1218{
1219	GList *list = NULL;
1220	char buf[32];
1221	bundle *b;
1222	int ret;
1223	int fd;
1224
1225	b = bundle_create();
1226	if (!b) {
1227		_E("Out of memory");
1228		return AUL_R_ENOMEM;
1229	}
1230
1231	snprintf(buf, sizeof(buf), "%d", uid);
1232	bundle_add(b, AUL_K_TARGET_UID, buf);
1233
1234	fd = aul_sock_send_bundle(AUL_UTIL_PID, uid, COMP_INFO_FOREACH,
1235			b, AUL_SOCK_ASYNC);
1236	bundle_free(b);
1237	if (fd < 0)
1238		return aul_error_convert(fd);
1239
1240	ret = aul_sock_recv_pkt_with_cb(fd, __comp_info_foreach_cb,
1241			&list);
1242	if (ret < 0) {
1243		g_list_free_full(list, __destroy_comp_info);
1244		return aul_error_convert(ret);
1245	}
1246
1247	*info_list = list;
1248
1249	return AUL_R_OK;
1250}
```

Reply:
yes.

Reply:
Really?
If multiple request are sent from mutiple thread, 
It will access to socket at the same time from each thread
Is it really thread safe??

https://www.google.co.kr/search?newwindow=1&source=hp&ei=SHJTXdqhHM2smAXM6Kb4CA&q=socket+read+thread+safe&oq=socket+read+thread+safe&gs_l=psy-ab.3..0i30j0i8i30l2.664.6662..6848...1.0..0.361.3713.0j19j4j2......0....1..gws-wiz.......35i39j0i10j0i131j0j0i203j0i5i30j33i160.TaZfdv-90N4&ved=0ahUKEwja5cjUqIHkAhVNFqYKHUy0CY8Q4dUDCAU&uact=5

Reply:
Each clients create a socket fd to access to the socket. And then, the clients send and receive messages using each fds.
Is it not thread safe? Could you explain what you are worried?
If I misunderstand, please tell me.

Reply:
On the same application, 
this API could be called on multiple thread
and it seems not thread safe

Reply:
Ok, i understand, per each request are create dedicated socket, and it used only per request
So, it seem to thread safe